### PR TITLE
Add SignedData func

### LINKED ISSUE #510: Add SignedData() to things that can be signed (have Sign())
## Is your feature request related to a problem? Please describe.
I'm always frustrated when I'm looking at #373. One can add some signature generate elsewhere, but he doesn't always have the data to sign since `Marshal()` is not the same as `SignedData()` (that exists for object session tokens, for example).

## Describe the solution you'd like
Add some methods to serialize data to sign.

## Describe alternatives you've considered
Not so much when we need completely detached signatures.

## Additional context
#373.

### PR DESCRIPTION
**Pull request by smallhive**
closes #510

Added func to known places with `Sign` method, plus for some places when I think it would be useful.
Of course, in some way it is almost similar to `Marshal` func, but for consistency called it _SignedData_.

If I missed any place, just mention it

### CODE REVIEW

File: container/container.go [hunk 1]
```diff
@@ -155,6 +155,11 @@ func (x Container) Marshal() []byte {
 	return x.v2.StableMarshal(nil)
 }
 
+// SignedData returns actual payload which would be signed, if you call [client.Client.ContainerPut] method.
```
Comment (roman-khimov) [inline container/container.go:158]: `... call [Container.CalculateSignature] ...`?

**Review by cthulhu-rider — COMMENTED** [review]
what about `object.Object`?

---

also suggest to add unified test to `crypto/test`:
```go
type SignedComponent interface {
	SignedData() []byte
	Sign(neofscrypto.Signer) error
	VerifySignature() bool
}

func TestSignedData(tb testing.TB, signer neofscrypto.Signer, cmp SignedComponent) {
	data := cmp.SignedData()
	
	sig, err := signer.Sign(data)
	require.NoError(tb, err)
	
	err = cmp.Sign(neofscrypto.NewStaticSigner(signer.Scheme(), sig, signer.Public()))
	require.NoError(tb, err)
	
	require.True(tb, cmp.VerifySignature())
}
```
and test this against all touched components

**Review by carpawell — COMMENTED** [review]
`SignedData returns data that _would be signed_` breaks me a little. Mabe `SignedData returns the data that is signed after a [Token.Sign] call`?

File: container/container.go (hunk 1 above)
Comment (smallhive) [reply container/container.go]: Updated

File: bearer/bearer.go [hunk 2]
```diff
@@ -268,6 +268,11 @@ func (b *Token) Sign(signer neofscrypto.Signer) error {
 	return nil
 }
 
+// SignedData returns actual payload which would be signed, if you call [Token.Sign] method.
```
[the remark this answers is not in the document]
Comment (smallhive) [reply bearer/bearer.go]: Updated

**Comment by smallhive** [comment]
> what about `object.Object`?

Nothing, it has a different API for signing with different names. I tend to think we may rewrite it to be compatible with `SignedComponent` interface. I did these changes in the last commit

I've added the suggested test, but we had some changes according to this

**Comment by cthulhu-rider** [comment]
> `SignedData returns data that _would be signed_` breaks me a little. Mabe `SignedData returns the data that is signed after a [Token.Sign] call`?

```
// SignedData returns signed data of the T.
//
// See also [T.Sign].
```
?

File: crypto/test/tests.go [hunk 3]
```diff
@@ -32,3 +32,25 @@ func RandomSignerRFC6979(tb testing.TB) user.Signer {
...
+
+// SignedComponent describes component which can signed and the signature may be verified.
+type SignedComponent interface {
+	SignedData() []byte
+	Sign(user.Signer) error
+	VerifySignature() bool
+}
+
+// SignedDataComponent tests [SignedComponent] for valid data generation by SignedData function.
+func SignedDataComponent(tb testing.TB, signer user.Signer, cmp SignedComponent) {
```
Comment (cthulhu-rider) [inline crypto/test/tests.go:44]: `signer` may be of type `neofscrypto.Signer`, isn't it?

File: bearer/bearer.go [hunk 4]
```diff
@@ -253,8 +253,8 @@ func (b Token) AssertUser(id user.ID) bool {
 // Note that any Token mutation is likely to break the signature, so it is
 // expected to be calculated as a final stage of Token formation.
 //
-// See also VerifySignature, Issuer.
-func (b *Token) Sign(signer neofscrypto.Signer) error {
+// See also VerifySignature, Issuer, [Token.SignedData].
+func (b *Token) Sign(signer user.Signer) error {
```
Comment (cthulhu-rider) [inline bearer/bearer.go:257]: dont do this just for test, make a workaround for the test if needed

File: crypto/test/tests.go (hunk 3 above)
Comment (smallhive) [reply crypto/test/tests.go:44]: No, it can't be. If it `neofscrypto.Signer` we cover just couple of structs

File: bearer/bearer.go (hunk 4 above)
Comment (smallhive) [reply bearer/bearer.go:257]: In this case we should duplicate `SignedComponent` for `neofscrypto.Signer` and for `user.Signer`

File: bearer/bearer.go [hunk 5]
```diff
@@ -253,7 +253,7 @@ func (b Token) AssertUser(id user.ID) bool {
 // Note that any Token mutation is likely to break the signature, so it is
 // expected to be calculated as a final stage of Token formation.
 //
-// See also VerifySignature, Issuer.
+// See also VerifySignature, Issuer, [Token.SignedData].
```
Comment (roman-khimov) [inline bearer/bearer.go:256]: Doc links should be fixed at the same time.

File: bearer/bearer.go [hunk 6]
```diff
@@ -268,6 +268,13 @@ func (b *Token) Sign(signer neofscrypto.Signer) error {
 	return nil
 }
 
+// SignedData returns actual payload for sign.
```
Comment (roman-khimov) [inline bearer/bearer.go:271]: `to sign`? I always remember @EliChin at these moments.

File: container/container.go [hunk 7]
```diff
@@ -467,7 +474,7 @@ func (x Container) ReadDomain() (res Domain) {
 // [neofscrypto.ECDSA_DETERMINISTIC_SHA256] scheme, for example, [neofsecdsa.SignerRFC6979]
 // can be used.
 //
-// See also VerifySignature.
+// See also VerifySignature, [Container.SignedData].
```
Comment (roman-khimov) [inline container/container.go:477]: Link (and elsewhere).

File: bearer/bearer.go (hunk 4 above)
Comment (roman-khimov) [reply bearer/bearer.go]: But why? This is certainly not worth a change for test. And tests should be fine with `neofscrypto.Signer` since `user.Signer` includes it.

File: crypto/test/tests.go [hunk 8]
```diff
@@ -32,3 +32,48 @@ func RandomSignerRFC6979(tb testing.TB) user.Signer {
 
 	return user.NewAutoIDSignerRFC6979(p.PrivateKey)
 }
+
+// SignedComponent describes component which can signed and the signature may be verified.
+type SignedComponent interface {
+	SignedData() []byte
+	Sign(neofscrypto.Signer) error
```
Comment (roman-khimov) [inline crypto/test/tests.go:39]: Likely some generic magic can be used here, but two interfaces are kinda OK for test as well.

File: object/fmt.go [hunk 9]
```diff
@@ -93,9 +93,10 @@ func (o *Object) VerifyID() error {
 	return nil
 }
 
-// CalculateAndSetSignature signs id with provided key and sets that signature to
-// the object.
-func (o *Object) CalculateAndSetSignature(signer neofscrypto.Signer) error {
+// Sign signs id with provided key and sets that signature to the object.
```
Comment (roman-khimov) [inline object/fmt.go:96]: `signs object ID`?

**Comment by smallhive** [comment]
Updated. Make two test interfaces/functions for each `user.Signer` and `neofscrypto.Signer`. But Object functions I renamed to be compatible with new test interface and more important with other signable components

**Comment by smallhive** [comment]
No problem with the signer it is OK. Problem with checking functions
```
func SignedDataComponent(tb testing.TB, signer neofscrypto.Signer, cmp SignedComponent)

func SignedDataComponentUser(tb testing.TB, signer user.Signer, cmp SignedComponentUserSigner)
```

They don't want to accept the wrong types for `cmp` parameter.

```
Cannot use 'val' (type *session.Container) as the type SignedComponent Type does not implement 'SignedComponent'
need the method: Sign(neofscrypto.Signer) error 
have the method: Sign(signer user.Signer) error
```

File: object/fmt.go (hunk 9 above)
Comment (smallhive) [reply object/fmt.go]: updated

File: container/container.go (hunk 7 above)
Comment (smallhive) [reply container/container.go]: Updated

File: bearer/bearer.go (hunk 6 above)
Comment (smallhive) [reply bearer/bearer.go]: updated

File: bearer/bearer.go (hunk 5 above)
Comment (smallhive) [reply bearer/bearer.go]: updated

**Comment by smallhive** [comment]
That is why we have two interfaces and two functions to check them

File: bearer/bearer.go [hunk 10]
```diff
@@ -244,16 +244,16 @@ func (b Token) AssertUser(id user.ID) bool {
...
-// Sign calculates and writes signature of the Token data using issuer's signer.
+// Sign calculates and writes signature of the [Token] data using issuer's signer.
 // Returns signature calculation errors.
 //
-// Sign MUST be called if Token is going to be transmitted over
+// Sign MUST be called if [Token] is going to be transmitted over
 // NeoFS API V2 protocol.
 //
-// Note that any Token mutation is likely to break the signature, so it is
+// Note that any [Token] mutation is likely to break the signature, so it is
```
Comment (roman-khimov) [inline bearer/bearer.go:253]: These are not related, of course. I meant only those lines that were changed. Can we separate these changes into a different commit?

Comment (smallhive) [reply bearer/bearer.go:253]: We can, but why not related? Updating documentation in the same place as `See also [Token.VerifySignature], [Token.Issuer], [Token.SignedData].` What the difference between this row and other documentation in this function?

Comment (roman-khimov) [reply bearer/bearer.go:253]: You're inevitably touching "See also [Token.VerifySignature], [Token.Issuer], [Token.SignedData]" in your PR, so you can change links as well. You don't need to touch these comments to deal with `SignedData`, that's why it's a separate change. Good one! But separate.

Comment (smallhive) [reply bearer/bearer.go:253]: OK, I see, I will separate these changes

Comment (smallhive) [reply bearer/bearer.go:253]: Moved to separate commit

**Review by roman-khimov — APPROVED** [review]

**Review by carpawell — APPROVED** [review]

**Comment by cthulhu-rider** [comment]
```
need the method: Sign(neofscrypto.Signer) error 
have the method: Sign(signer user.Signer) error
```
needed method is more generic and may be implemented in-place

File: crypto/test/tests.go (hunk 3 above)
Comment (cthulhu-rider) [reply crypto/test/tests.go]: it's not about structs, it's about function body. It deals with `user.Signer` as `neofscrypto.Signer`

File: crypto/test/tests.go [hunk 11]
```diff
@@ -32,3 +32,48 @@ func RandomSignerRFC6979(tb testing.TB) user.Signer {
...
+// SignedDataComponentUser tests [SignedComponentUserSigner] for valid data generation by SignedData function.
+func SignedDataComponentUser(tb testing.TB, signer user.Signer, cmp SignedComponentUserSigner) {
+	data := cmp.SignedData()
+
+	sig, err := signer.Sign(data)
+	require.NoError(tb, err)
+
+	static := neofscrypto.NewStaticSigner(signer.Scheme(), sig, signer.Public())
+
+	err = cmp.Sign(user.NewSigner(static, signer.UserID()))
```
Comment (cthulhu-rider) [inline crypto/test/tests.go:75]: does `VerifySignature` check passed user? If not, then `SignedDataComponent` would be enough (we may provide fast cast from `SignedComponentUserSigner` to `SignedComponent`

Comment (roman-khimov) [reply crypto/test/tests.go:75]: I see what you mean with wrapping, but it's a test code and it can be improved later.
